lib/block_view_validator.go

```diff
@@ -0,0 +1,2133 @@
+	if operationData.Type != OperationTypeRegisterAsValidator {
+		return fmt.Errorf(
+			"_disconnectRegisterAsValidator: trying to revert %v but found %v",
+			OperationTypeRegisterAsValidator,
+			operationData.Type,
+		)
+	}
+
+	// Convert TransactorPublicKey to TransactorPKID.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(currentTxn.PublicKey)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return errors.Wrapf(RuleErrorInvalidValidatorPKID, "_disconnectRegisterAsValidator: ")
+	}
+
+	// Delete the current ValidatorEntry.
+	currentValidatorEntry, err := bav.GetValidatorByPKID(transactorPKIDEntry.PKID)
+	if err != nil {
+		return errors.Wrapf(err, "_disconnectRegisterAsValidator: ")
+	}
+	// Note that we don't need to check isDeleted because the Get returns nil if isDeleted=true.
+	if currentValidatorEntry == nil {
+		return fmt.Errorf(
+			"_disconnectRegisterAsValidator: no ValidatorEntry found for %v", transactorPKIDEntry.PKID,
+		)
+	}
+	bav._deleteValidatorEntryMappings(currentValidatorEntry)
+
+	// Restore the PrevValidatorEntry, if exists. The PrevValidatorEntry won't exist if this was
+	// the first time this ValidatorEntry was created. The PrevValidatorEntry will exist if this
+	// was an update operation on an existing ValidatorEntry.
+	prevValidatorEntry := operationData.PrevValidatorEntry
+	if prevValidatorEntry != nil {
+		bav._setValidatorEntryMappings(prevValidatorEntry)
+	}
+
+	// Disconnect the BasicTransfer.
+	return bav._disconnectBasicTransfer(
+		currentTxn, txHash, utxoOpsForTxn[:operationIndex], blockHeight,
+	)
+}
+
+func (bav *UtxoView) _connectUnregisterAsValidator(
+	txn *MsgDeSoTxn,
+	txHash *BlockHash,
+	blockHeight uint32,
+	verifySignatures bool,
+) (
+	_totalInput uint64,
+	_totalOutput uint64,
+	_utxoOps []*UtxoOperation,
+	_err error,
+) {
+	// Validate the starting block height.
+	if blockHeight < bav.Params.ForkHeights.ProofOfStakeNewTxnTypesBlockHeight {
+		return 0, 0, nil, errors.Wrapf(RuleErrorProofofStakeTxnBeforeBlockHeight, "_connectUnregisterAsValidator: ")
+	}
+
+	// Validate the txn TxnType.
+	if txn.TxnMeta.GetTxnType() != TxnTypeUnregisterAsValidator {
+		return 0, 0, nil, fmt.Errorf(
+			"_connectUnregisterAsValidator: called with bad TxnType %s", txn.TxnMeta.GetTxnType().String(),
+		)
+	}
+
+	// Connect a basic transfer to get the total input and the
+	// total output without considering the txn metadata.
+	totalInput, totalOutput, utxoOpsForTxn, err := bav._connectBasicTransfer(
+		txn, txHash, blockHeight, verifySignatures,
+	)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectUnregisterAsValidator: ")
+	}
+	if verifySignatures {
+		// _connectBasicTransfer has already checked that the txn is signed
+		// by the top-level public key, which we take to be the sender's
+		// public key so there is no need to verify anything further.
+	}
+
+	// Validate the transactor.
+	if err = bav.IsValidUnregisterAsValidatorMetadata(txn.PublicKey); err != nil {
```

Comment:
I notice that we allow a validator to Unregister even if they're currently jailed. This seems to create a loophole where I can Unregister in order to escape jail sooner than I otherwise could.

Imagine the following scenario:
- I'm jailed for 48 epochs, but unregistering can be done in 3 epochs.
- I staked to myself, which means that it's no trouble for me to Unregister (versus if I had a lot of other people staked to me)
- If the above is true, then I could Unregister and Re-register to escape jail.

I think a reasonable thing would be to disallow Unregister if the validator is jailed by adding the check to IsValidUnregisterAsvalidatorMetadata()